http-client/src/main/java/io/micronaut/http/client/DefaultHttpClient.java

```diff
@@ -1796,18 +1796,43 @@ private String getHostHeader(URI requestURI) {
 
             @Override
             protected void channelRead0(ChannelHandlerContext channelHandlerContext, FullHttpResponse fullResponse) {
-                HttpResponseStatus status = fullResponse.status();
-                int statusCode = status.code();
-                HttpStatus httpStatus;
                 try {
-                    httpStatus = HttpStatus.valueOf(statusCode);
-                } catch (IllegalArgumentException e) {
-                    emitter.onError(e);
-                    return;
+
+                    HttpResponseStatus status = fullResponse.status();
+                    int statusCode = status.code();
+                    HttpStatus httpStatus;
+                    try {
+                        httpStatus = HttpStatus.valueOf(statusCode);
+                    } catch (IllegalArgumentException e) {
+                        if (complete.compareAndSet(false, true)) {
+                            emitter.onError(e);
+                        } else {
+                            getLog().warn("unsupported http status after completion", e);
+                        }
+                        return;
+                    }
+
+                    handleResponse(fullResponse, status, statusCode, httpStatus);
+                } finally {
+                    if (fullResponse.refCnt() > 0) {
+                        try {
+                            ReferenceCountUtil.release(fullResponse);
+                        } catch (Throwable e) {
+                            if (LOG.isDebugEnabled()) {
+                                LOG.debug("Failed to release response: {}", fullResponse);
+                            }
+                        }
+                    }
+                    if (!HttpUtil.isKeepAlive(fullResponse)) {
+                        keepAlive = false;
+                    }
+                    pipeline.remove(this);
+
                 }
+            }
 
+            private void handleResponse(FullHttpResponse fullResponse, HttpResponseStatus status, int statusCode, HttpStatus httpStatus) {
```

Comment:
Rather than extracting this out into its own method, just move the status handling inside the `try`

Reply:
OK - it means that there is now conflict with master, but no big deal.